File: 2_quicksort_recursive.py
def quicksort_recursive(arr):
    """Recursive quicksort implementation"""
    if len(arr) <= 1:
        return arr

    pivot = arr[len(arr) // 2]
    left = [x for x in arr if x < pivot]
    middle = [x for x in arr if x == pivot]
    right = [x for x in arr if x > pivot]

    return quicksort_recursive(left) + middle + quicksort_recursive(right)

# Test the recursive implementation
arr = [64, 34, 25, 12, 22, 11, 90, 88, 45, 50]

print("=== Quicksort with Recursion ===")
print(f"Original array: {arr}")

sorted_arr = quicksort_recursive(arr)

print(f"Sorted array: {sorted_arr}")

File: 2_quicksort_recursive.c
#include <stdio.h>

// Helper function to partition the array
int partition(int arr[], int low, int high) {
    int pivot = arr[high];
    int i = low - 1;

    for (int j = low; j < high; j++) {
        if (arr[j] < pivot) {
            i++;
            int temp = arr[i];
            arr[i] = arr[j];
            arr[j] = temp;
        }
    }
    int temp = arr[i + 1];
    arr[i + 1] = arr[high];
    arr[high] = temp;
    return i + 1;
}

// Recursive quicksort
void quicksort_recursive(int arr[], int low, int high) {
    if (low < high) {
        int pi = partition(arr, low, high);
        quicksort_recursive(arr, low, pi - 1);   // Sort left partition
        quicksort_recursive(arr, pi + 1, high);   // Sort right partition
    }
}

// Print array
void print_array(int arr[], int n) {
    for (int i = 0; i < n; i++) {
        printf("%d ", arr[i]);
    }
    printf("\n");
}

int main() {
    int arr[] = {64, 34, 25, 12, 22, 11, 90, 88, 45, 50};
    int n = sizeof(arr) / sizeof(arr[0]);

    printf("=== Quicksort with Recursion ===\n");
    printf("Original array: ");
    print_array(arr, n);

    quicksort_recursive(arr, 0, n - 1);

    printf("Sorted array: ");
    print_array(arr, n);

    return 0;
}
